escrow/lib.rs

```diff
@@ -0,0 +1,254 @@
+//! can be non-negligible on some chains.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+pub mod escrow {
+    use scale::{Decode, Encode};
+
+    #[ink(storage)]
+    pub struct Escrow {
+        /// Buyer's account
+        buyer: AccountId,
+
+        /// Sellers's account
+        seller: AccountId,
+
+        /// Deposit value
+        deposit: Balance,
+    }
+
+    /// Error scenarios in escrow contract
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// Someone else than buyer tries to confirm
+        ConfirmNotAsBuyer,
+
+        /// Someone else than seller tries to refund
+        RefundNotAsSeller,
+
+        /// requested transfer failed, this can be the case if the contract does not
+        /// have sufficient free funds or if the transfer would have brought the
+        /// contract's balance below minimum balance
+        TransferFailed,
+    }
+
+    /// An event emitted when token transfer occurs
+    #[ink(event)]
+    pub struct Transfer {
+        #[ink(topic)]
+        to: AccountId,
+        value: Balance,
+    }
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl Escrow {
+        /// Instantiates new escrow contract with buyer as contract author
+        #[ink(constructor, payable)]
+        pub fn new(seller: AccountId) -> Self {
+            Self {
+                buyer: Self::env().caller(),
+                seller,
+                deposit: Self::env().transferred_value(),
+            }
+        }
+
+        /// Returns currently stored deposit
+        #[ink(message)]
+        pub fn get_deposit(&self) -> Balance {
+            self.deposit
+        }
+
+        /// Returns buyers account
+        #[ink(message)]
+        pub fn get_buyer(&self) -> AccountId {
+            self.buyer
+        }
+
+        /// Returns buyers account
+        #[ink(message)]
+        pub fn get_seller(&self) -> AccountId {
+            self.seller
+        }
+
+        /// The buyer confirms delivery. Contract transfers a deposit to the seller and terminates itself
+        #[ink(message)]
+        pub fn confirm(&mut self) -> Result<()> {
+            let caller = Self::env().caller();
+            if caller != self.buyer {
+                return Err(Error::ConfirmNotAsBuyer);
+            }
+            self.make_transfer(self.seller, self.deposit)?;
+            self.env().terminate_contract(caller)
+        }
+
+        /// The seller aborts transaction. Contract refunds a deposit to the buyer and terminates itself
+        #[ink(message)]
+        pub fn refund(&mut self) -> Result<()> {
+            let caller = Self::env().caller();
+            if caller != self.seller {
+                return Err(Error::RefundNotAsSeller);
+            }
+            self.make_transfer(self.buyer, self.deposit)?;
+            self.env().terminate_contract(caller)
```

Comment:
Optional: if transferring the deposit for a deleted contract to the seller instead of the buyer bothers you, here you can just  do `terminate_contract(self.buyer)`

Reply:
While `terminate_contract(self.buyer)` indeed transfers _free_  remaining contract's balance always to `self.buyer`, it turns out it transfers contract's instantiation storage deposit to caller anyway.